crates/ibc/src/services/core/context.rs

```diff
@@ -0,0 +1,93 @@
+use crate::core::ics02_client::client_state::Status;
+use crate::core::ics02_client::client_type::ClientType;
+use crate::core::ics03_connection::connection::IdentifiedConnectionEnd;
+use crate::core::ics04_channel::channel::IdentifiedChannelEnd;
+use crate::core::ics04_channel::packet::Sequence;
+use crate::core::ics24_host::identifier::{ClientId, ConnectionId};
+use crate::core::ics24_host::path::{AckPath, ChannelEndPath, CommitmentPath, Path};
+use crate::core::{ContextError, ValidationContext};
+use crate::prelude::*;
+use crate::Height;
+
+/// Context to be implemented by the host to provide proofs in gRPC query responses
+///
+/// Trait used for the [`gRPC query services`](crate::services).
+pub trait ProvableContext {
+    /// Returns the proof for the given path at the given height.
+    /// As this is in the context of IBC, the path is expected to be an [`IbcPath`](Path).
+    fn get_proof(&self, height: Height, path: &Path) -> Option<Vec<u8>>;
+}
+
+/// Context to be implemented by the host that provides gRPC query services.
+///
+/// Trait used for the [`gRPC query services`](crate::services).
+pub trait QueryContext: ProvableContext + ValidationContext {
+    // Client queries
+
+    /// Returns the list of all clients.
+    fn client_states(
+        &self,
+    ) -> Result<Vec<(ClientId, <Self as ValidationContext>::AnyClientState)>, ContextError>;
+
+    /// Returns the list of all consensus states for the given client.
+    fn consensus_states(
+        &self,
+        client_id: &ClientId,
+    ) -> Result<Vec<(Height, <Self as ValidationContext>::AnyConsensusState)>, ContextError>;
+
+    /// Returns the list of all heights at which consensus states for the given client are.
+    fn consensus_state_heights(&self, client_id: &ClientId) -> Result<Vec<Height>, ContextError>;
+
+    /// Returns the status of the given client.
+    fn client_status(&self, client_id: &ClientId) -> Result<Status, ContextError>;
+
+    /// Returns the list of supported client types.
+    fn allowed_clients(&self) -> Vec<ClientType>;
```

Comment:
In principle, the list of allowed clients should come from the own IBC validation/execution contexts. Though, currently, this feature is missing, and hosts cannot manage the list of allowed clients(get or set client params like [here](https://github.com/cosmos/ibc-go/blob/5ca37ef6e56a98683cf2b3b1570619dc9b322977/modules/core/02-client/keeper/keeper.go#L423-L441) in ibc-go). Thus, any implementation for this method would force hosts to stick with it forever.

I suggest we feature this capability in a separate PR first or alternatively we can add a comment acknowledging this issue and temporarily returning `None`` for client params. Nonetheless, the get/set params APIs should be featured soon and the implementation of this method get updated accordingly. WDYT?

Reply:
I see. I would suggest returning `None` and adding a _TODO_ in the method. Also, removing `allowed_clients` from `QueryContext` trait.

Reply:
I don't think we need this method (note: it is in `ValidationContext` at the time of writing). Checking if a client is "allowed" is implicitly done [here](https://github.com/cosmos/ibc-rs/blob/133863db5f9ce56d57b2502913a21bde0fad7bda/crates/ibc/src/core/ics02_client/handler/create_client.rs#L42). If the host doesn't support a given client type, `ctx.client_state()` will fail.

What was the use of this method when in `QueryContext`? If you need it there, then I don't see the problem with having it live there.

Reply:
- `.allowed_clients()` is needed for client params query.
- The way you explained the implicit validation of ClientType via `ctx.client_state(&client_id).is_ok()`, I disagree with the returned error `ClientError::ClientStateAlreadyExists`.

Reply:
My mistake; the [call](https://github.com/cosmos/ibc-rs/blob/133863db5f9ce56d57b2502913a21bde0fad7bda/crates/ibc/src/core/ics02_client/handler/create_client.rs#L28) to `decode_client_state()` would fail

Reply:
The way I see it, it'd be nice if we could somehow get it from `ClientState` derive macro. Do you agree?

Reply:
As far as I see, the primary purpose of the `allowed_client` is not for checking which `ClientState` an ibc-rs implementation supports. It’s about enabling hosts to update client params (as `ibc-go` does [here](https://github.com/cosmos/ibc-go/blob/5ca37ef6e56a98683cf2b3b1570619dc9b322977/modules/core/keeper/msg_server.go#L703-L713)). Without it, hosts can’t add or remove allowed `ClientType`, which typically comes from governance decisions (or generally from a decentralized authority as hosts may adopt different approaches in this regard)

Reply:
I am confused, since currently the `ValidationContext::allowed_clients()` is only used in core handlers, and nowhere else. So I don't see the link with queries?

Reply:
@plafer, sorry for the confusion. I was working on a possible solution in which I changed some code in the final PR. I will revert them now.

The discussion about `allowed_clients` seems nuanced. So, I will drop this method and return unimplemented status on ClientParamsQuery in this PR. Let's deal with module params support in future PRs.